### PR TITLE
Add __next__ return

### PR DESCRIPTION
**Pull request by davidhewitt**
This builds on #996 by enabling ability to `return` a value in the `StopIteration` exception.

If you like this approach I'll write some docs in the guide before we merge this PR.

### CODE REVIEW

File: src/class/iter.rs [hunk 1]
```diff
@@ -64,22 +64,37 @@ impl PyIterMethods {
     }
 }
 
-pub struct IterNextOutput(Option<PyObject>);
+pub enum IterNextOutput {
+    Yield(PyObject),
+    Return(Option<PyObject>),
+}
```
Comment (kngwyu) [inline src/class/iter.rs:70]: I suggest
```rust
pub enum IterNextOutput<T, U> {
    Yield(T),
    Return(U),
}

pub type PyIterNextOutput = IterNextOutput<PyObject, PyObject>;

impl IntoPyCallbackOutput<*mut ffi::PyObject> for PyIterNextOutput {
    fn convert(self, _py: Python) -> PyResult<*mut ffi::PyObject> {
        match self {
            IterNextOutput::Yield(o) => Ok(o.into_ptr()),
            IterNextOutput::Return(opt) => Err(crate::exceptions::StopIteration::py_err((opt,))),
        }
    }
}

impl<T, U> IntoPyCallbackOutput<PyIterNextOutput> for IterNextOutput<T, U>
where
    T: IntoPy<PyObject>,
    U: IntoPy<PyObject>,
{
    fn convert(self, py: Python) -> PyResult<PyIterNextOutput> {
        match self {
            IterNextOutput::Yield(o) => Ok(IterNextOutput::Yield(o.into_py(py))),
            IterNextOutput::Return(o) => Ok(IterNextOutput::Return(o.into_py(py))),
        }
    }
}

impl<T> IntoPyCallbackOutput<PyIterNextOutput> for Option<T>
where
    T: IntoPy<PyObject>,
{
    fn convert(self, py: Python) -> PyResult<PyIterNextOutput> {
        match self {
            Some(o) => Ok(PyIterNextOutput::Yield(o.into_py(py))),
            None => Ok(PyIterNextOutput::Return(py.None())),
        }
    }
}
```

**Comment by kngwyu** [comment]
Thank you! Looks good but I want to make general `IterNextOutput<T, U>` work.

**Comment by davidhewitt** [comment]
Thanks, I like that design. I've pushed that to the branch and added a small sample to the guide.

File: guide/src/class.md [hunk 2]
```diff
@@ -891,6 +890,52 @@ impl PyIterProtocol for Container {
 For more details on Python's iteration protocols, check out [the "Iterator Types" section of the library
 documentation](https://docs.python.org/3/library/stdtypes.html#iterator-types).
 
+#### Returning a value from iteration
```
Comment (kngwyu) [inline guide/src/class.md:893]: I think `PyIterProtocols`'s doc would be a better place for this document. `class.md` is too long now.

File: guide/src/class.md [hunk 3]
```diff
@@ -891,6 +890,52 @@ impl PyIterProtocol for Container {
...
+for example with the following code:
+
+```python
+def iter_():
+    for i in range(5):
+        yield i
+    return "Ended"
+```
+
+When the generator `iter_()` returns with `"Ended"`, this is translated to a `StopIteration` exception `e` with `e.value = "Ended"`.
```
Comment (kngwyu) [inline guide/src/class.md:905]: This is an implementation detail and I don't think it's practically important.

**Comment by davidhewitt** [comment]
Good feedback, thanks - I've shortened the doc and moved most of it to `PyIterProtocol`.

**Comment by kngwyu** [comment]
Looks rustfmt complains...

File: guide/src/class.md [hunk 4]
```diff
@@ -891,6 +890,14 @@ impl PyIterProtocol for Container {
 For more details on Python's iteration protocols, check out [the "Iterator Types" section of the library
 documentation](https://docs.python.org/3/library/stdtypes.html#iterator-types).
 
+#### Returning a value from iteration
+
+This guide has so far shown how to use `Option<T>` to implement yielding values during iteration.
+In Python a generator can also return a value. To express this in Rust, PyO3 provides the
+[`IterNextOutput`](https://docs.rs/pyo3/latest/pyo3/class/iter/enum.IterNextOutput.html) enum to
```
Comment (kngwyu) [inline guide/src/class.md:897]: Nice :+1:

**Comment by davidhewitt** [comment]
🤦 pushed fixed formatting.

**Comment by kngwyu** [comment]
Thanks!

**Pull request merged**